nexus/src/app/certificate.rs

```diff
@@ -0,0 +1,230 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! x.509 Certificates
+
+use crate::context::OpContext;
+use crate::db;
+use crate::db::lookup;
+use crate::db::lookup::LookupPath;
+use crate::db::model::Name;
+use crate::external_api::params;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::DataPageParams;
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::NameOrId;
+use openssl::pkey::PKey;
+use openssl::x509::X509;
+use ref_cast::RefCast;
+use std::sync::Arc;
+use uuid::Uuid;
+
+fn validate_certs(input: Vec<u8>) -> Result<(), String> {
+    let certs = X509::stack_from_pem(&input.as_slice())
+        .map_err(|err| format!("Failed to parse certificate as PEM: {err}"))?;
+    if certs.is_empty() {
+        return Err("could not parse".to_string());
+    }
+    Ok(())
+}
+
+fn validate_private_key(key: Vec<u8>) -> Result<(), String> {
+    let _ = PKey::private_key_from_pem(&key.as_slice())
+        .map_err(|err| format!("Failed to parse private key as PEM: {err}"))?;
+
+    Ok(())
+}
+
+impl super::Nexus {
+    pub fn certificate_lookup<'a>(
+        &'a self,
+        opctx: &'a OpContext,
+        certificate: &'a NameOrId,
+    ) -> lookup::Certificate<'a> {
+        match certificate {
+            NameOrId::Id(id) => {
+                LookupPath::new(opctx, &self.db_datastore).certificate_id(*id)
+            }
+            NameOrId::Name(name) => LookupPath::new(opctx, &self.db_datastore)
+                .certificate_name(Name::ref_cast(name)),
+        }
+    }
+
+    pub async fn certificate_create(
```

Comment:
Some doc comments would be useful here...I was surprised to see that this only seems to handle Nexus certificates.  Maybe even rename these to `nexus_certificate_*()`?  (not a big deal for sure)

Reply:
Updated in 184811fde2594c7d4fc52a4de8cb4728a4457c58 - I'm adding the service type in the API. I expected we'd handle certificates for services other than Nexus eventually, but had originally punted it. I can include it now (as an enum of one variant), to make the intent more clear.